### PR TITLE
Implementing load/save legacy for both grid and scattered locations

### PR DESCRIPTION
**Pull request by exepulveda**
Implementing load/save legacy for both grid and scattered locations.

Now the module has load/save _legacy that manages scattered and grid datasets, which are fully compatible with all GSLIB programs.

### CODE REVIEW

**Comment by juliohm** [comment]
Awesome. I'll try to take a look by the end of the week. +1

**Review by juliohm — CHANGES_REQUESTED** [review]
Thank you for the proposed improvements.

General comments to help in future contributions:

1. Only add a dependency in a PR if strictly needed. The addition of DataFrames.jl here was unnecessary for example.
2. Changes to working code should be structured. In most cases, when existing working code is modified to a new version, the changes need to be minimal with a clear indication of why they were needed. Your changes to the existing load_legacy function are quite breaking for no clear reason and are also hard to follow/review. You are changing the logic of the function body, and doing disruptive changes that we will hardly be able to merge without extensive testing. Ideally, you would concentrate your attention to a specific portion of the function body that needs generalization, and then would write the modifications with small additions or auxiliary functions like the parser functions you suggested.
3. Formatting fixes should come either as a separate PR or as a final step after the core functionality has been reviewed and tested. Right now we have many changes to the tests which are unrelated to the proposed functionality. Reviewers have to spend time reading the changes to make sure that no typo was introduced, etc etc. The more effort you put into facilitating the lives of the reviewers, the faster the changes will get merged.

File: src/GslibIO.jl [hunk 1]
```diff
@@ -45,35 +49,98 @@ function load(file::File{format"GSLIB"})
...
     # properties names
-    vars = [Symbol(readline(fs)) for i in 1:nvars]
+    vars = [strip(readline(fs)) for i in 1:nvars]
 
     # read property values
     X = readdlm(fs)
 
-    # handle missing values
-    replace!(X, na=>NaN)
+    return X, header, nvars, extrainfo, vars
```
[the remark this answers is not in the document]
Comment (exepulveda) [reply src/GslibIO.jl:75]: Not sure if the statement you have made has been proven. When you parse something that has N components and you return those N components, that is totally a good design. Maybe you are concern about returning a tuple instead a structure. A structure will be one thing but semantically will be the same N-thing.

I'm afraid that parsing must return these 5 things (that is what a GSLIB format is). I could create a non-mutable structure if that helps.

The "return" thing is easy to fix :-)

File: src/GslibIO.jl [hunk 2]
```diff
@@ -45,35 +49,98 @@ function load(file::File{format"GSLIB"})
...
+    coordprops = coordnames
+  elseif eltype(coordnames) === Int
+    for var in coordnames
+      @assert 1 <= var <= length(propnames) "$var is not a variable in the GSLIB file"
+    end
+    coordprops = [propnames[i] for i in coordnames] # transform the properties from index to Symbol
+  else
+    itemtype = eltype(coordnames)
+    error("Coordinates must be an iterable of String, Symbol or Integer, but not '$itemtype'")
   end
```
[the remark this answers is not in the document]
Comment (exepulveda) [reply src/GslibIO.jl:135]: Maybe we should for clarity, but we will be changing the API to current users. So, what should I do?

File: test/runtests.jl [hunk 3]
```diff
@@ -35,17 +35,54 @@ datadir = joinpath(@__DIR__,"data")
...
+    sdata_grid = GslibIO.load_legacy(fname, (2, 2, 2))
+    @test size(domain(sdata_grid)) == (2, 2, 2)
+    @test origin(domain(sdata_grid)) == [0., 0., 0.]
+    @test spacing(domain(sdata_grid)) == [1., 1., 1.]
+    por = sdata_grid[:Porosity]
+    lit = sdata_grid[:Lithology]
+    sat = sdata_grid[Symbol("Water Saturation")]
+    @test isequal(por, [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8])
+    @test isequal(lit, [1, 2, 3, 4, 5, 6, 7, 8])
+    @test isequal(sat, [0.8, 0.7, 0.8, 0.7, 0.8, 0.7, 0.8, NaN])
```
[the remark this answers is not in the document]
Comment (exepulveda) [reply test/runtests.jl:49]: Well, I though that the space in function call is the convention, therefore I assumed that it was the perfect time to make the changes. If we don't, do you want that the new code does not used a white space to avoid mixing formats?

File: src/GslibIO.jl [hunk 4]
```diff
@@ -45,35 +49,98 @@ function load(file::File{format"GSLIB"})
...
 
     # properties names
-    vars = [Symbol(readline(fs)) for i in 1:nvars]
+    vars = [strip(readline(fs)) for i in 1:nvars]
 
     # read property values
     X = readdlm(fs)
 
-    # handle missing values
-    replace!(X, na=>NaN)
```
[the remark this answers is not in the document]
Comment (exepulveda) [reply src/GslibIO.jl:69]: Simply because the parser should not take such action. The parse just parses the content as the file is.
You should be able to see the missing values handled in lines 94 and 134 respectively, which can be improved to avoid duplication of code. 

Remember that the previous code only could parse grid (and with a bug). To make the load/save generic is worthy to have the parser as a separated function.

File: src/GslibIO.jl [hunk 5]
```diff
@@ -45,35 +49,98 @@ function load(file::File{format"GSLIB"})
...
-    readline(fs)
+    header = readline(fs)
 
-    # number of properties
-    nvars = parse.(Int, readline(fs))
+    # number of properties is the first integer (it may contain extra info)
+    line = readline(fs)
+    linecont = split(line)
+    nvars = parse(Int, linecont[1])
+    extrainfo = linecont[1:end]
```
[the remark this answers is not in the document]
Comment (exepulveda) [reply src/GslibIO.jl:67]: I think you don't know well the GSLIB format. In GSLIB there is not two different things for grid or scattered locations. Both have the same format. The only difference is that for grids, no coordinates as variables are needed as they can be inferred from the grid definition, and depending on the version, the second line may have the grid definition as an extra (but optional and unused) information.

Basically a GSLIB file has
a) The header in the first line (not used indeed)
b) The number of variables in the second line (plus extra information but not used in practice)
c) The name of the variables in the following lines
d) The data: each line will have the values according to the number of variables in b) delimited by spaces

And that is what exactly you should be able to see. Indeed the new code is now capable to parse grid and pset with no code replication. 

If you could elaborate more about refactoring the code to be more self-contained, I could be able to follow what you want. It is hard to guess what you have in mind, but I doing my best.

File: src/GslibIO.jl [hunk 6]
```diff
@@ -45,35 +49,98 @@ function load(file::File{format"GSLIB"})
...
+    error("Coordinates must be an iterable of String, Symbol or Integer, but not '$itemtype'")
   end
+
+  # handle missing values
+  replace!(X, na=>NaN)
+
+  # create data dictionary
+  data = (; zip(propnames, eachcol(X))...)
+
+  georef(DataFrame(data), coordprops)
```
[the remark this answers is not in the document]
Comment (exepulveda) [reply src/GslibIO.jl:143]: I believe you, but I will check this :-)

File: src/GslibIO.jl [hunk 7]
```diff
@@ -7,12 +7,16 @@ module GslibIO
...
+using DataFrames
 
 using GeoStatsBase
 
 # type aliases
 const Array2or3{T} = Union{AbstractArray{T,2},AbstractArray{T,3}}
 
+# For naming coordinates in a grid
+const DefaultCoordNames = [:x, :y, :z]
+
```
[the remark this answers is not in the document]
Comment (exepulveda) [reply src/GslibIO.jl:19]: No existing function is using :x, :y or :z, because you are working with grid and you don't need to store coordinates.
They are needed when you store a pset for naming the variables that are coordinates. Maybe I can define it inside the function instead.

[the remark this answers is not in the document]
Comment (juliohm) [reply src/GslibIO.jl:19]: Yes, that would be a great. Just pass these names as a default option instead of creating an external variable :+1:

File: src/GslibIO.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (juliohm) [reply src/GslibIO.jl:67]: Could you please write a single `parse_gslib` function that consumes the text file and returns a struct with predefined and well-documented fields? After we have a clear shared vision of this struct, and we can post-process it to convert to either a grid or a pset.

Ultimately what we want is to construct a spatial domain (grid or pset) and a table with values for the actual covariates.

[the remark this answers is not in the document]
Comment (juliohm) [reply src/GslibIO.jl:67]: Let's start with a simple `parse_gslib` function. As soon as this function is ready and tested, we can continue working on the PR to use this function with minimal modifications.

File: src/GslibIO.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (juliohm) [reply src/GslibIO.jl:75]: Ultimately we want to return two high-level concepts: a spatial domain and a table with values. Any intermediate constructs are low-level details that could be wrapped in higher level concepts. This makes the code easier to understand and maintain in the long run. Alternatively, create a struct with 5 well-documented fields and we will be able to review it and see if a different struct would be preferred that is more aligned with Julia designs out there.

File: src/GslibIO.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (juliohm) [reply src/GslibIO.jl:69]: I like this. Fully agree that the parser should just consume the raw data and that we need to post-process it later to convert missing values for example. I am looking forward to seeing this clean `parse_gslib` function before we can continue reviewing the other modifications proposed here. Remember that this function should return a well-documented struct with the GSLIB spec. I still think that the spec is different for point set and grid, so I am looking forward to learn from you this single spec.

File: src/GslibIO.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (juliohm) [reply src/GslibIO.jl:135]: I don't understand how the refactoring would change the external API for users. They will still use the same `load/save` and `load_legacy/save_legacy` without bothering about how the code is written internally.

File: test/runtests.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (juliohm) [reply test/runtests.jl:49]: You tend to mix up changes in a pull request that have different nature. It is fine to do so after other members have reviewed your work, but showing a big diff with whitespace changes isn't very helpful for reviewers who want to make sure that the PR does not introduce bugs.

So if you spot these inconsistencies, we will be happy to fix them. But this PR is about a major change to the code to support point sets. It is already a lot of changes to review. Can you see why reviewers would be skeptical about such changes here? You need to put yourself in the side of the reviewer always. Imagine that you have a project that is out there for a while and someone submits a PR with a lot of unrelated changes. Besides the difficulty to review the work, we also mix up commits of different nature and this is hard to track in the future or to revert in case a bug is found.

[the remark this answers is not in the document]
Comment (exepulveda) [reply test/runtests.jl:49]: You haven't answer my question. Should I use or not white space in the new code given the current code does not use white space?

[the remark this answers is not in the document]
Comment (juliohm) [reply test/runtests.jl:49]: In the new code you can always be consistent with the conventions adopted. The issue is in the old code in the tests that were changed. Strictly speaking we can always hide whitespace changes in git diffs, but my comment about reversing commits with bugs still holds. If someone wants to revert your commit, they will also revert the whitespace changes unintentionally. We want to avoid such things.

So in summary: new code can be written according to the convention. wrong conventions in old code can be fixed with separate commits, preferably in separate PRs.

**Comment by exepulveda** [comment]
Hi @juliohm,

I have rolled back the code to just address parse_gslib and facilitate your review. This change only introduces parse_gslib function which is now used by load_legacy.

After this new commit is hopefully merged, I will overload load_legacy (or a new function) for the case of generic GSLIB file. Note that a generic GSLIB does not necessarily have to contain locations. 

Regards

File: src/GslibIO.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (exepulveda) [reply src/GslibIO.jl]: If you rename load_legacy to parse_grid the API will change.

File: src/GslibIO.jl [hunk 8]
```diff
@@ -43,6 +51,38 @@ function load(file::File{format"GSLIB"})
...
+"""
+    parse_gslib(filename::AbstractString)
+
+Helper function to parse a standard GSLIB file.
+It returns the data as a matrix `X`, and `metadata` that includes the parsed information.
+"""
+
+function parse_gslib(filename::AbstractString)
+  open(filename) do fs
+    # header
```
Comment (juliohm) [inline src/GslibIO.jl:63]: notice that this comment isn't very helpful

File: src/GslibIO.jl [hunk 9]
```diff
@@ -43,6 +51,38 @@ function load(file::File{format"GSLIB"})
...
+    header = readline(fs)
+    # number of variables is the first integer in the second line (it may contain extra info)
```
Comment (juliohm) [inline src/GslibIO.jl:65]: this comment is not very helpful either. this type of information should live in the struct definition. here you just need to read the spec following the name conventions adopted in the struct and people will understand clearly what is happening.

File: src/GslibIO.jl [hunk 10]
```diff
@@ -43,6 +51,38 @@ function load(file::File{format"GSLIB"})
...
+    line = readline(fs)
+    linesplit = split(line)
+    nvars = parse(Int, linesplit[1])
+    if length(linesplit) > 1
+      loc = findfirst(linesplit[1], line)
+      extrainfo = lstrip(line[loc.stop+1:end]) # extrainfo is kept as it is after `nvars`
+    else
+      extrainfo = "" # empty in case extrainfo is not provided
+    end
+    # properties names
```
Comment (juliohm) [inline src/GslibIO.jl:75]: please remove this comment as well

File: src/GslibIO.jl [hunk 11]
```diff
@@ -43,6 +51,38 @@ function load(file::File{format"GSLIB"})
...
+    header = readline(fs)
+    # number of variables is the first integer in the second line (it may contain extra info)
+    line = readline(fs)
+    linesplit = split(line)
+    nvars = parse(Int, linesplit[1])
+    if length(linesplit) > 1
+      loc = findfirst(linesplit[1], line)
+      extrainfo = lstrip(line[loc.stop+1:end]) # extrainfo is kept as it is after `nvars`
+    else
+      extrainfo = "" # empty in case extrainfo is not provided
```
Comment (juliohm) [inline src/GslibIO.jl:73]: please remove this comment, it is easily inferred from the context

File: src/GslibIO.jl [hunk 12]
```diff
@@ -43,6 +51,38 @@ function load(file::File{format"GSLIB"})
...
+    end
+    # properties names
+    vars = [strip(readline(fs)) for i in 1:nvars]
+
+    metadata = GSLIBMetadata(header, nvars, vars, extrainfo)
+
+    # read data
+    data = readdlm(fs)
+
+    data, metadata
```
Comment (juliohm) [inline src/GslibIO.jl:83]: So you read the spec and the data table, but then return in the opposite order. Please call the variables `meta` and `data` and return `meta, data`.

File: src/GslibIO.jl [hunk 13]
```diff
@@ -52,28 +92,20 @@ Optionally set the value used for missings `na`.
 """
 function load_legacy(filename::AbstractString, dims::NTuple{3,Int};
                      origin=(0.,0.,0.), spacing=(1.,1.,1.), na=-999)
-  open(filename) do fs
-    # skip header
-    readline(fs)
 
-    # number of properties
-    nvars = parse.(Int, readline(fs))
+  # parse the GSLIB file
```
Comment (juliohm) [inline src/GslibIO.jl:96]: this comment isn't very helpful

File: src/GslibIO.jl [hunk 14]
```diff
@@ -52,28 +92,20 @@ Optionally set the value used for missings `na`.
...
+  X, metadata = parse_gslib(filename)
```
Comment (juliohm) [inline src/GslibIO.jl:97]: Please choose a single identifier and stick to it. Here you are renaming `data` to `X`. The reviewer of your contribution has to maintain two different identifiers in his/her head.

File: src/GslibIO.jl [hunk 15]
```diff
@@ -52,28 +92,20 @@ Optionally set the value used for missings `na`.
...
-    readline(fs)
 
-    # number of properties
-    nvars = parse.(Int, readline(fs))
+  # parse the GSLIB file
+  X, metadata = parse_gslib(filename)                     
 
-    # properties names
-    vars = [Symbol(readline(fs)) for i in 1:nvars]
+  vars = [Symbol(v) for v in metadata.varnames]
```
Comment (juliohm) [inline src/GslibIO.jl:99]: I don't think this variable is needed. One can use `meta.varnames` directly?

File: src/GslibIO.jl [hunk 16]
```diff
@@ -52,28 +92,20 @@ Optionally set the value used for missings `na`.
...
-    # read property values
-    X = readdlm(fs)
+  # handle missing values
+  replace!(X, na=>NaN)
 
-    # handle missing values
-    replace!(X, na=>NaN)
+  # create data dictionary
+  data = (; zip(vars, eachcol(X))...)
+  domain = RegularGrid(dims, origin, spacing)
```
Comment (juliohm) [inline src/GslibIO.jl:106]: So it seems to me that you are not using the coordinates in the point set case?

File: test/data/legacy_scatter.gslib [hunk 17]
```diff
@@ -0,0 +1,9 @@
+Legacy GSLIB file
```
Comment (juliohm) [inline test/data/legacy_scatter.gslib:1]: Rename this file to `legacy_pset.gslib` and adjust the header like you did with the grid case:

```
# Legacy GSLIB point set file
```

**Review by juliohm — CHANGES_REQUESTED** [review]

File: src/GslibIO.jl [hunk 18]
```diff
@@ -13,6 +13,14 @@ using GeoStatsBase
 # type aliases
 const Array2or3{T} = Union{AbstractArray{T,2},AbstractArray{T,3}}
 
+# structure for GSLIB metadata
+struct GSLIBMetadata
```
[the remark this answers is not in the document]
Comment (exepulveda) [reply src/GslibIO.jl:17]: Is Spec for Specification?

File: src/GslibIO.jl [hunk 19]
```diff
@@ -13,6 +13,14 @@ using GeoStatsBase
...
+  header    # Content of the first line (documentation only)
+  nvars     # An integer with the number of variables (second line)
+  varnames  # The variable names in the following `nvars` lines
+  extrainfo # This corresponds to the extra info in the second line after `nvars`
```
[the remark this answers is not in the document]
Comment (exepulveda) [reply src/GslibIO.jl:21]: There is not wrong/right order as the structure is high level. I placed it at the end because it is optional and not actually used. I think it is much better to keep it at the end.

File: src/GslibIO.jl (hunk 15 above)
Comment (exepulveda) [reply src/GslibIO.jl:99]: I need to check if the variable names need to be Symbols or just Strings.

File: src/GslibIO.jl (hunk 16 above)
Comment (exepulveda) [reply src/GslibIO.jl:106]: You have missed the comment of the commit. The idea is to first have the load_legacy working with parse_legacy for a easier review. When this is done, I will submit other PR with load_legacy for point sets.

File: src/GslibIO.jl (hunk 18 above)
[the remark this answers is not in the document]
Comment (exepulveda) [reply src/GslibIO.jl:17]: It is :-)

File: src/GslibIO.jl (hunk 14 above)
Comment (exepulveda) [reply src/GslibIO.jl:97]: I did not quite understand you comment. The current code already has a data variable used in line 99 and was using the X variable. That is the reason to keep X as variable.

File: src/GslibIO.jl (hunk 15 above)
Comment (exepulveda) [reply src/GslibIO.jl:99]: It seems line 99 (which is part the current code) only works with a array of Symbols. I could inject ```[Symbol(v) for v in metadata.varnames]``` in line 99 to avoid the vars variable, although in my view is clearer to use that variable.

File: src/GslibIO.jl (hunk 19 above)
[the remark this answers is not in the document]
Comment (exepulveda) [reply src/GslibIO.jl:21]: > So this extrainfo contains what exactly? Since it is coming immediately after the `nvars`, I would place it in the right order in the struct.

I think the documentation is very clear. I will give you an example if you are not familirised with GSLIB format.
If you krige a regular grid using kt3d, the output will contain two variables (estimation and variance), but some versions of GSLIB add after the second line meant for the number of variables (2 is this case) the grid specification (usually just number of nodes, but sometimes nodes, starts and sizes). This extra information is just for annotate the grid specification, but is never used in any GSLIB program. Therefore, you will see something like:
```
KT3D ESTIMATES WITH:  samples_au                                                
    2   10   20   10
Estimate
EstimationVariance
0.62118578     0.81496090    
```
In this example, nvars is 2, whereas `extrainfo` will be "10   20   10".

`extrainfo` is not used, even we could not include it in the structure, but I think it is better to parse everything in case users want to do something with it.

I hope this explanation helps you to better understand the GSLIB format. You could read the official doc here: http://www.gslib.com/gslib_help/format.html.

File: src/GslibIO.jl (hunk 14 above)
Comment (juliohm) [reply src/GslibIO.jl:97]: You can rename this `X` variable to `data` and use the same identifier inside the `parse_legacy` function.

File: src/GslibIO.jl (hunk 15 above)
Comment (juliohm) [reply src/GslibIO.jl:99]: They need to be symbols. I would parse them as symbols inside the `parse_legacy` function.

File: src/GslibIO.jl (hunk 16 above)
Comment (juliohm) [reply src/GslibIO.jl:106]: I see. Please keep working on this PR, we don't need a separate PR. After the suggestions above are incorporated we can take a second look in the code, and generalize to the point set case.

File: src/GslibIO.jl (hunk 19 above)
[the remark this answers is not in the document]
Comment (juliohm) [reply src/GslibIO.jl:21]: Based on your explanation, the `extrainfo` field should be an *arbitrary* string that is ignored in most applications. I agree we should remove it from the struct. If someone complains about it or has a use case in the future that requires parsing this extra info, we can always submit a new PR to fix the issue. Other lines are also allowed to have comments like the lines with variable names, but we don't want to save comments in general.

File: src/GslibIO.jl (hunk 15 above)
Comment (exepulveda) [reply src/GslibIO.jl:99]: > They need to be symbols. I would parse them as symbols inside the `parse_legacy` function.

I prefer not to do it in the `parse_legacy` function, because the parser should be low level (variables names are just Strings). The use of Symbols is after parsing.

Comment (juliohm) [reply src/GslibIO.jl:99]: Do we gain something from this string representation? I mean, are we planning to do any string processing? If not, please just convert the types to `Integer`, `String`, `Float64` etc. This will make the code clearer. If this was a major specification it would make more sense to save the raw strings, but here we only have a tiny header.

File: src/GslibIO.jl (hunk 14 above)
Comment (exepulveda) [reply src/GslibIO.jl:97]: But then the variable `data` will be replaced by something else (not good).

Comment (exepulveda) [reply src/GslibIO.jl:97]: I will use datadict for the variable name after `# create data dictionary`

Comment (juliohm) [reply src/GslibIO.jl:97]: Please use `data` as suggested, the identifier `X` isn't very good in this context. Also, to be consistent with the current view of the project, we look at these as tables as opposed to dictionaries. You can use either `data` or `table` for this identifier. It should be clear from the context.

**Comment by exepulveda** [comment]
I have a problem: if table is a NnamedTuple with the data parsed, and coordnames is the tuple with the variables that are coordinates, it seems there is not way to use table directly but convert it first to DataFrace.
```
table = (x = rand(10), y = rand(10), cu = rand(10))
coordnames = (:x, :y)
```
The next code fails
```
georef(table, coordnames)
ERROR: MethodError: georef(::NamedTuple{(:x, :y, :cu),Tuple{Array{Float64,1},Array{Float64,1},Array{Float64,1}}}, ::Tuple{Symbol,Symbol}) is ambiguous. Candidates:
  georef(table, coordnames::Tuple{Vararg{T,N}} where T where N) in GeoStatsBase at /home/a1634120/.julia/packages/GeoStatsBase/4zuvI/src/georef.jl:24
  georef(tuple::NamedTuple, domain) in GeoStatsBase at /home/a1634120/.julia/packages/GeoStatsBase/4zuvI/src/georef.jl:40
Possible fix, define
  georef(::NamedTuple, ::Tuple{Vararg{T,N}} where T where N)
```
but this works fine
```
georef(DataFrame(table), coordnames)
```

I may need to use DataFrames anyway :-)

**Comment by juliohm** [comment]
You don't need a DataFrame, take a look at the georef methods here: https://juliaearth.github.io/GeoStats.jl/stable/data

Maybe we should add a new method for NamedTuple + Symbol.

**Comment by exepulveda** [comment]
> You don't need a DataFrame, take a look at the georef methods here: https://juliaearth.github.io/GeoStats.jl/stable/data
> 
> Maybe we should add a new method for NamedTuple + Symbol.

None of the current methods can handle NamedTuple + Symbol. There are two options:
a) To use DataFrame to call `georef(DataFrame(table), coordnames)`
b) Add to GeoStatsBase a new function `georef(table, coordnames) = georef(DataFrame(table), coordnames)`

Which would you prefer?

**Comment by juliohm** [comment]
Please create a domain and georef the named tuple to it. The domain can be
a RegularGrid or PointSet

**Comment by juliohm** [comment]
Since you are not understanding how the georef function works, let me clarify a bit further. Your named tuple should only contain the actual variables, not the columns with coordinates. Then you can georef this named tuple to any kind of domain.

**Comment by juliohm** [comment]
@exepulveda anything I can do to clarify the issues further? I found this link with some good advise on code contributions and reviews: https://mtlynch.io/code-review-love

**Comment by exepulveda** [comment]
> @exepulveda anything I can do to clarify the issues further?

Yes. I still don't understand why the simplest and clearest piece of code
```
georef(DataFrame(table), coordnames)
```
is discouraged. If you can explain it to me with enough details that will be great.

**Comment by exepulveda** [comment]
> @exepulveda I found this link with some good advise on code contributions and reviews: https://mtlynch.io/code-review-love

I will check it. Thanks.

**Comment by juliohm** [comment]
The reason why we don't want to use `DataFrame(table)` explicitly is because it requires adding DataFrames.jl as an explicit dependency in GslibIO.jl. In the documentation, we already have alternative methods that are preferred and that will make the code more general. A DataFrame is a specific type of table with more features than what is strictly needed to implement the GeoStats.jl functionality. The goal is to eliminate occurrences of `DataFrame` from the codebase whenever possible. At the end we only need to comply to the Tables.jl api and users will be able to manipulate SQL databases, and other kinds of tables that are not a simple in-memory DataFrame.

**Comment by exepulveda** [comment]
> The reason why we don't want to use `DataFrame(table)` explicitly is because it requires adding DataFrames.jl as an explicit dependency in GslibIO.jl. In the documentation, we already have alternative methods that are preferred and that will make the code more general. A DataFrame is a specific type of table with more features than what is strictly needed to implement the GeoStats.jl functionality. The goal is to eliminate occurrences of `DataFrame` from the codebase whenever possible. At the end we only need to comply to the Tables.jl api and users will be able to manipulate SQL databases, and other kinds of tables that are not a simple in-memory DataFrame.

On the other hand, there is a function that accepts DataFrame as input; if the APi offers to me this why not to use it. Beware that there will a cost of duplicating code, because I will have to split parsed variables into the ones that are coordinates from the ones that are attributes. That is already done in the current implementation with DataFrames. Nevertheless, if you want that I do it in that way, that is OK.

**Comment by juliohm** [comment]
Yes, as I pointed out, the ultimate goal is to get rid of DataFrames.jl completely. It is an unnecessary dependency in GeoStats.jl that is there only because we didn't have the time to rewrite some algorithms with general Tables.jl tables. So introducing another occurrence of DataFrame here to be removed in the future is not very helpful nor aligned with this end goal.

I suggest that you write a pseudo-code first with the algorithm, and only then start writing code. We can review the pseudo-code quickly in a language close to English, and then put time and effort writing/reviewing actual code. This will also make the review process more engaging. Here is my attempt to write such pseudo-code with the knowledge I have of the legacy GSLIB format:

```julia
# the format comes in two flavors: grid and pset
# the grid case is already covered in the current
# implementation, and consists of reading the
# header of the file with grid specs, and a table
# containing just variables to be georeferenced
# in this grid, here is how the parser works:

1. read grid spec from header
2. read table with variables
3. georef(table, grid)

# the other flavor is the pset flavor and we lack
# an implementation for it, so here is an attempt
# to describe the format, and corresponding parser
# in the case of a pset, we have a table of values
# with columns of different kind, namely variables
# and coordinates in a coordinate system
# we need to read the header with the variable
# names, read the table with the variables and
# coordinates and then subtract the coordinates
# from the table before georeferencing it to a pset

1. read variables names from header
2. read table with variables and coordinates
3. extract columns with coordinates to build pset
4. extract columns with variables to build table
5. georef(table, pset)
```

See how the plan is very clear? It helps the reviewer understand what you have in your mind, and helps both of us catch misinterpretations of the format and possibly of the current project functionality.

**Comment by exepulveda** [comment]
This code without DataFrames works well:
```
  data, metadata = parse_legacy(filename)
  # create data dictionary
  tableall = Dict(metadata.varnames, eachcol(data))

  # create data for coordinates
  coords = transpose(reduce(hcat, [tableall[c] for c in coordnames]))
  # create table with varnames not in coordnames
  attrnames = [c for c in metadata.varnames if c ∉  coordnames]
  table = (; zip(attrnames, [tableall[c] for c in attrnames])...)
  georef(table, PointSet(coords))
```

Any further suggestion to apply before a new commit?

**Comment by exepulveda** [comment]
> the ultimate goal is to get rid of DataFrames.jl completely

Sorry if I did not get this before, but I think first time you mention this explicitly (I might be wrong). If you had mentioned this before I would had understood the reason not to use DataFrames in the first place.

I guess the code above is exactly the pseudo code that we use to ask young programmers. I think we are now close :-)

**Comment by juliohm** [comment]
> Any further suggestion to apply before a new commit?

Feel free to commit. It is easier to review the code when it is in the PR because I can select each individual line and provide better feedback.

> Sorry if I did not get this before, but I think first time you mention this explicitly (I might be wrong). If you had mentioned this before I would had understood the reason not to use DataFrames in the first place.

I don't remember if I mentioned it somewhere else, but I had provided a review here saying that we didn't want DataFrames.jl as a dependency and you kept insisting on it as if my review was not based on something relevant. When I provide a review in a PR you need to assume that I have a clear vision for this project that I've been developing for almost 4 years now.

**Comment by exepulveda** [comment]
> I don't remember if I mentioned it somewhere else, but I had provided a review here saying that we didn't want DataFrames.jl as a dependency and you kept insisting on it as if my review was not based on something relevant. When I provide a review in a PR you need to assume that I have a clear vision for this project that I've been developing for almost 4 years now.

You are missing the opportunity to explain this clear vision for this project to the project's collaborators. If I don't understand something I will ask. Just executing instructions without explanations is not good for collaboration. If you explain more I understand better for next contributions to this project.

Maybe PRs are not the best place to discuss this, but where to discuss these matters?

**Comment by juliohm** [comment]
Asking is totally fine and I encourage you to do so. My concern is not about you asking questions, it is about a general attitude you've been showing in PRs --- an attitude that is triggered whenever I suggest something that you can't comprehend because you don't have the full view of the project as I do. You tend to reply with a "I prefer it my way" attitude and instead of fixing these minor issues that would be trivial to fix on your side, we spend a lot of time discussing details that aren't very interesting to discuss. If you can read that link I shared above on good PRs and reviews, that will already help a lot, and we will be able to focus on what matters the most moving forward.

We can always chat over Zulip or Gitter or any other channel to clarify things, I am always available. I think it would be a good idea to setup regular online meetings in 2021 with all contributors. That way we can get to know each other and can share what we are currently working on. I will organize this event. The major challenge is to find a time of the day that is good for everyone in different time zones.

**Comment by exepulveda** [comment]
> My concern is not about you asking questions, it is about a general attitude you've been showing in PRs --- an attitude that is triggered whenever I suggest something that you can't comprehend because you don't have the full view of the project as I do. You tend to reply with a "I prefer it my way" attitude and instead of fixing these minor issues that would be trivial to fix on your side, we spend a lot of time discussing details that aren't very interesting to discuss. If you can read that link I shared above on good PRs and reviews, that will already help a lot, and we will be able to focus on what matters the most moving forward.

Let's move on and let's try to find a right moment to discuss this because I also feel some negative attitude to me (I mentioned this to you before). It may be useful for you to listen the point of view of a contributor, both of us can learn from each other. Reading the blog article you sent me is in my TODO list.

**Comment by exepulveda** [comment]
> I think it would be a good idea to setup regular online meetings in 2021 with all contributors

It is a terrific idea.

**Comment by exepulveda** [comment]
Last commit addresses, I believe, what you have requested. I hope you can check it soon.

File: src/GslibIO.jl [hunk 20]
```diff
@@ -13,6 +13,14 @@ using GeoStatsBase
 # type aliases
 const Array2or3{T} = Union{AbstractArray{T,2},AbstractArray{T,3}}
 
+# legacy specification of the GSLIB format
+# (see documentation here: http://www.gslib.com/gslib_help/format.html)
+struct LegacySpec
+  header::String          # Content of the first line (documentation only)
+  nvars::Int              # An integer with the number of variables (second line)
+  varnames::Array{Symbol} # The variable names as Symbol in the following `nvars` lines
```
Comment (juliohm) [inline src/GslibIO.jl:21]: I would make it a `Vector{Symbol}` to be even more specific.

File: src/GslibIO.jl (hunk 9 above)
Comment (juliohm) [reply src/GslibIO.jl:59]: this comment has not been addressed, yet you marked it as resolved

File: src/GslibIO.jl [hunk 21]
```diff
@@ -43,6 +51,38 @@ function load(file::File{format"GSLIB"})
...
+It returns the data as a matrix `X`, and `metadata` that includes the parsed information.
+"""
+
+function parse_gslib(filename::AbstractString)
+  open(filename) do fs
+    # header
+    header = readline(fs)
+    # number of variables is the first integer in the second line (it may contain extra info)
+    line = readline(fs)
+    linesplit = split(line)
```
[the remark this answers is not in the document]
Comment (juliohm) [reply src/GslibIO.jl:61]: The proposed change hasn't been considered.

File: src/GslibIO.jl (hunk 10 above)
Comment (juliohm) [reply src/GslibIO.jl:63]: another comment marked as resolved inappropriately

**Review by juliohm — CHANGES_REQUESTED** [review]
If you check your PR you will see that it has the same issues I pointed out before, namely:

1. It ignores many comments and suggestions I made. You marked comments as resolved without fixing nor replying to them.
2. You are implementing various different things in this PR without enough experience in code reviews. So you are again trying to provide a mix of features (load_legacy + save_legacy + LegacySpec + ...) and I am having a hard time to review it.

Please fix the comments that are still open, undo your changes related to `save_legacy`, and also the changes you made to the README.md. Keep it simple until we are able to efficiently communicate modifications in the code, and are feeling safe about merging them.

**Comment by exepulveda** [comment]
I think is better not to continue with this PR until I understand better how you want to operate (sometimes is quite confusing) and I read/follow better the best practices for contributing to. At the end I was only writing a Notebook for comparing GSLIB and GeoStats :-)
